packages/flutter_tools/lib/src/commands/build_appbundle.dart

```diff
@@ -77,11 +95,146 @@ class BuildAppBundleCommand extends BuildSubCommand {
     return usage;
   }
 
+  // TODO(garyq): Support non-hardcoded prebuild for setup verification.
+  /// Hardcoded environment to perform deferred component setup with.
+  Environment createEnvironment() {
+    final FlutterProject flutterProject = FlutterProject.current();
+    String output = flutterProject.directory
+        .childDirectory('build')
+        .childDirectory('app')
+        .childDirectory('intermediates')
+        .childDirectory('flutter')
+        .childDirectory('release').path;
+    final Environment result = Environment(
+      outputDir: globals.fs.directory(output),
+      buildDir: flutterProject.directory
+          .childDirectory('.dart_tool')
+          .childDirectory('flutter_build'),
+      projectDir: flutterProject.directory,
+      defines: <String, String>{ kSplitAot: 'true', kBuildMode: 'release', kTargetPlatform: 'android-arm64' },
+      inputs: <String, String>{},
+      cacheDir: globals.cache.getRoot(),
+      flutterRootDir: globals.fs.directory(Cache.flutterRoot),
+      artifacts: globals.artifacts,
+      fileSystem: globals.fs,
+      logger: globals.logger,
+      processManager: globals.processManager,
+      engineVersion: globals.artifacts.isLocalEngine
+        ? null
+        : globals.flutterVersion.engineRevision
+    );
+    return result;
+  }
+
   @override
   Future<FlutterCommandResult> runCommand() async {
     if (globals.androidSdk == null) {
       exitWithNoSdkMessage();
     }
+
+    // TODO(garyq): Refactor out to a better place.
+    // Deferred components prebuild setup verification.
+    List<DeferredComponent> deferredComponents = FlutterProject.current().manifest.deferredComponents;
+    bool deferredComponentsEnabled = boolArg('deferred-components') && deferredComponents != null;
+    if (deferredComponentsEnabled) {
```

Comment:
This command should not be running the build twice, instead have the build itself try and fail with the detailed messages. Not that you can't do some light validation here too.

Reply:
I split the setup validation into two sections. The first section handles validating the existence of android modules and files, which do not require knowing loading units. The second section runs after assemble, and handles all loading unit related checks.